lpd/utils/threshold_checker.py

```diff
@@ -0,0 +1,62 @@
+from abc import ABC, abstractmethod
+from typing import Union
+
+from torch import Tensor
+
+
+class ThresholdChecker(ABC):
+    """
+    Check if the current value is better than the previous best value according to different threshold criteria
+    This is an abstract class meant to be inherited by different threshold checkers
+    Can also be inherited by the user to create custom threshold checkers
+    """
+    def __init__(self, threshold: float):
+        self.threshold = threshold
+
+    @abstractmethod
+    def is_new_value_higher(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        pass
+
+    @abstractmethod
+    def is_new_value_lower(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        pass
+
+
+class AbsoluteThresholdChecker(ThresholdChecker):
+    """
+    A threshold checker that checks if the difference between the current value and the previous best value
+     is greater than or equal to a given threshold
+
+    Args:
+        threshold - the threshold to check (must be non-negative)
+    """
+    def __init__(self, threshold: float = 0.0):
+        if threshold < 0:
+            raise ValueError(f"Threshold must be non-negative, but got {threshold}")
+        super(AbsoluteThresholdChecker, self).__init__(threshold)
+
+    def is_new_value_higher(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        return new_value - old_value > self.threshold
+
+    def is_new_value_lower(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        return old_value - new_value > self.threshold
+
+
+class RelativeThresholdChecker(ThresholdChecker):
+    """
+    A threshold checker that checks if the relative difference between the current value and the previous best value
+     is greater than or equal to a given threshold
+
+    Args:
+        threshold - the threshold to check (must be non-negative)
+    """
+    def __init__(self, threshold: float = 0.0):
+        if threshold < 0:
+            raise ValueError(f"Threshold must be non-negative, but got {threshold}")
+        super(RelativeThresholdChecker, self).__init__(threshold)
+
+    def is_new_value_higher(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        return (new_value - old_value) / old_value > self.threshold
+
+    def is_new_value_lower(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        return (old_value - new_value) / old_value > self.threshold
```

Comment:
looks like a bug
should be 
(old_value - new_value) / new_value
say 
threshold = 0.05 (5%)
old_value = 10
new_value = 9.51
a drop of 5% from 10 will give 10/1.05 = ~9.52
yet, (old_value - new_value) / old_value = (10-9.51)/10 = 0.049
which is lower than 0.05, but the threshold should be met in this case

the fix, (old_value - new_value) / new_value will yield:
(10-9.51)/9.51 = 0.0515 which is greater than the threshold, as expected

Reply:
So if I understand correctly, you are saying that the 5% change should be according to the new value?
This isn't how I was thinking about this method. The way I see it, if you lose 5% out of 10 you get 9.5. This means that in order to pass a 5% threshold you need to "lose" more value (in the is_new_value_lower case). So I believe that your example shouldn't meet the threshold requirement 😅

Reply:
got ya, sorry for the confusion